### PR TITLE
Additional bgs error handling in WarmBgsCacheJob

### PR DESCRIPTION
**Pull request by rubaiyat22**
Resolves [sentry](https://sentry.prod.appeals.va.gov/department-of-veterans-affairs/caseflow/issues/12781/?referrer=webhooks_plugin) and follows https://github.com/department-of-veterans-affairs/caseflow/pull/15668

- fix a typo in calling method `capture_exception`
- rescue all StandardError 
- update old tests

### CODE REVIEW

File: app/jobs/warm_bgs_caches_job.rb [hunk 1]
```diff
@@ -140,11 +140,14 @@ def warm_poa_and_cache_for_legacy_appeals(legacy_appeals, start_time, datadog_se
...
   def warm_poa_and_cache_for_ama_appeals(claimants, start_time, datadog_segment)
     appeals_to_cache = claimants.map do |claimant|
-      bgs_poa = claimant.power_of_attorney
-      claimant.update!(updated_at: Time.zone.now)
+      bgs_poa = claimant_poa(claimant)
+      next unless bgs_poa
 
       # [{:appeal_id=>1, :appeal_type=>"Appeal", :power_of_attorney_name=>"Clarence Darrow"}, ...]
-      warm_bgs_poa_and_return_cache_data(bgs_poa, claimant.decision_review_id, Appeal.name)
+      cache_data = warm_bgs_poa_and_return_cache_data(bgs_poa, claimant.decision_review_id, Appeal.name)
```
Comment (rubaiyat22) [inline app/jobs/warm_bgs_caches_job.rb:147]: do not update claimant if there's an error with BGS

File: app/jobs/warm_bgs_caches_job.rb [hunk 2]
```diff
@@ -172,7 +175,7 @@ def fetch_bgs_power_of_attorney_by_file_number(file_number, appeal_id)
     BgsPowerOfAttorney.find_or_create_by_file_number(file_number)
   rescue ActiveRecord::RecordInvalid # not found at BGS
     BgsPowerOfAttorney.new(file_number: file_number)
-  rescue Errno::ECONNRESET, Savon::HTTPError => error
+  rescue StandardError, BGS::ShareError => error
```
Comment (rubaiyat22) [inline app/jobs/warm_bgs_caches_job.rb:178]: rescue from ShareError

File: app/jobs/warm_bgs_caches_job.rb [hunk 3]
```diff
@@ -212,13 +215,20 @@ def claimants_for_open_appeals
 
   def oldest_claimants_with_poa
     oldest_claimants_for_open_appeals.limit(LIMITS[:OLDEST_CLAIMANT])
-      .select { |claimant| claimant.power_of_attorney.present? }
+      .select { |claimant| claimant_poa(claimant).present? }
```
Comment (rubaiyat22) [inline app/jobs/warm_bgs_caches_job.rb:218]: use new method `claimant_poa` which rescues errors

File: app/jobs/warm_bgs_caches_job.rb (hunk 2 above)
Comment (nanotone) [reply app/jobs/warm_bgs_caches_job.rb:178]: In my Rails console,
```ruby
> BGS::ShareError < StandardError
=> true
```

so this will have the same behavior if you leave out the `BGS::ShareError`

Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb:178]: ah my bad, i'll remove this now!

Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb]: done

File: app/jobs/warm_bgs_caches_job.rb (hunk 1 above)
Comment (nanotone) [reply app/jobs/warm_bgs_caches_job.rb:147]: I might be missing some context here, but is the claimant's `updated_at` column supposed to capture when we last successfully fetched their POA data? Because Rails uses that column to capture when the claimant record was last edited.

File: app/jobs/warm_bgs_caches_job.rb [hunk 4]
```diff
@@ -212,13 +215,20 @@ def claimants_for_open_appeals
...
 
   def oldest_claimants_for_open_appeals
     claimants_for_open_appeals.order(updated_at: :asc)
   end
 
+  def claimant_poa(claimant)
+    claimant.power_of_attorney
+  rescue StandardError => error
+    warning_msgs << "#{Appeal.name} #{claimant.decision_review_id}: #{error}"
+    nil
```
Comment (nanotone) [inline app/jobs/warm_bgs_caches_job.rb:229]: This is a cool method! I would be careful of using `nil` to represent that there was an error, because `claimant.power_of_attorney` can also return `nil` if BGS says the claimant doesn't have a POA. Assuming you care about the difference (maybe you don't?) some of this rescue/nil logic might have to be incorporated back into `warm_poa_and_cache_for_ama_appeals`.

File: app/jobs/warm_bgs_caches_job.rb (hunk 1 above)
Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb:147]: yes the former. It's being used to determine whether we should re-fetch the poa. For all new claimants that come into the system, we should fetch their POA first.

Comment (nanotone) [reply app/jobs/warm_bgs_caches_job.rb:147]: Circling back to this observation, now that I think I have a better sense of the PR as a whole.

It seems that you do want to skip `warm_bgs_poa_and_return_cache_data` even if `claimant_poa` ran successfully and just returned nil. But maybe you also want touch the `updated_at` column in that case? (Depending on how that column should be interpreted.) In which case the order and logic here might need a little reshuffling.

**Review by nanotone — COMMENTED** [review]
Good work! I left some questions and concerns in-line, but lemme know if they're hard to follow.

File: app/jobs/warm_bgs_caches_job.rb (hunk 1 above)
Comment (nanotone) [reply app/jobs/warm_bgs_caches_job.rb:147]: Ugh, :octocat: didn't show me the 3:00pm ET comment in time.

If you need to determine whether POA needs (re)fetching, I wonder if `claimant.updated_at` might be unreliable, since Rails will automatically touch it when changing any other field on the claimant record. (Currently I don't think we do, but we likely will in the future).

There is actually a place that almost fulfills your need: `bgs_power_of_attorneys.last_synced_at` -- however, it's in the other table, and if the claimant has no POA there's no place to store the `last_synced_at`. But it might be okay to say, any claimant with no POA could use a re-fetch?

File: app/jobs/warm_bgs_caches_job.rb (hunk 4 above)
Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb:229]: I created that method to ensure that the job can continue to warm and cache everything. I can make comments to clarify that i'm rescuing errors to let the job continue and it does not mean that claimant does not have a POA. I'm not seeing a need to consider the difference between no poa and error because we will just skip this claimant if it returns error.

File: app/jobs/warm_bgs_caches_job.rb (hunk 1 above)
Comment (lomky) [reply app/jobs/warm_bgs_caches_job.rb:147]: I concur with Yang's assessment - when it exists we should use the associated objects `last_synced_at` rather than rely on a circumstantially-trustworthy `updated_at`. Is there any downside to using BGS POAs field?

File: app/jobs/warm_bgs_caches_job.rb [hunk 5]
```diff
@@ -116,7 +116,7 @@ def warm_poa_for_oldest_cached_records
     oldest_bgs_poa_records.limit(LIMITS[:OLDEST_CACHED]).each do |bgs_poa|
       begin
         bgs_poa.save_with_updated_bgs_record! if bgs_poa.stale_attributes?
-      rescue Errno::ECONNRESET, Savon::HTTPError
+      rescue StandardError
         # no nothing
```
Comment (yoomlam) [inline app/jobs/warm_bgs_caches_job.rb:120]:
```suggestion
        # do nothing
```

We were doing nothing for those 2 specific errors that we know about. If you're generalizing the rescue, should we log it in case there are other error types:
```ruby
      rescue Errno::ECONNRESET, Savon::HTTPError
        # do nothing
      rescue StandardError => error
        capture_exception(error: error)
```

File: app/jobs/warm_bgs_caches_job.rb [hunk 6]
```diff
@@ -187,13 +190,13 @@ def warm_bgs_poa_and_return_cache_data(bgs_poa, appeal_id, appeal_type)
...
-  rescue Errno::ECONNRESET, Savon::HTTPError => error
-    warning_msgs << "#{appeal_type} #{appeal_id}: #{error}" if warning_msgs.count < 100
+  rescue StandardError => error
+    warning_msgs << "#{appeal_type} #{appeal_id}: #{error}"
     nil
   end
 
   def log_warning
-    slack_msg = warning_msgs.join("\n")
+    slack_msg = warning_msgs.first(100).join("\n")
```
Comment (yoomlam) [inline app/jobs/warm_bgs_caches_job.rb:199]: Why store all warning but only show the first 100?

**Review by yoomlam — COMMENTED** [review]
Some questions and suggestions

File: app/jobs/warm_bgs_caches_job.rb [hunk 7]
```diff
@@ -212,13 +215,20 @@ def claimants_for_open_appeals
...
     oldest_claimants_for_open_appeals.limit(LIMITS[:OLDEST_CLAIMANT])
-      .select { |claimant| claimant.power_of_attorney.present? }
+      .select { |claimant| claimant_poa(claimant).present? }
   end
 
   def oldest_claimants_for_open_appeals
     claimants_for_open_appeals.order(updated_at: :asc)
   end
 
+  def claimant_poa(claimant)
```
Comment (yoomlam) [inline app/jobs/warm_bgs_caches_job.rb:225]: I concur with Yang regarding returning nil when an error occurs. Errors should be handled by the method caller because it knows how to handle the error in its situation. 
Let's rename this method to make it clear that it will handle errors; something like:
```suggestion
  def claimant_poa_or_nil(claimant)
```

File: app/jobs/warm_bgs_caches_job.rb [hunk 8]
```diff
@@ -172,28 +172,26 @@ def fetch_bgs_power_of_attorney_by_file_number(file_number, appeal_id)
...
-        appeal_id: appeal_id,
-        appeal_type: appeal_type,
-        power_of_attorney_name: bgs_poa&.representative_name
-      }
-    end
-  rescue Errno::ECONNRESET, Savon::HTTPError => error
-    warning_msgs << "#{appeal_type} #{appeal_id}: #{error}" if warning_msgs.count < 100
+    bgs_poa.save_with_updated_bgs_record! if bgs_poa&.stale_attributes?
+
+    {
```
Comment (rubaiyat22) [inline app/jobs/warm_bgs_caches_job.rb:183]: Modify this so that the nil poa_name is cached in case poa changes to no poa

File: app/jobs/warm_bgs_caches_job.rb (hunk 7 above)
Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb]: update this

File: app/jobs/warm_bgs_caches_job.rb (hunk 6 above)
Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb]: i update this to only store 100

File: app/jobs/warm_bgs_caches_job.rb (hunk 5 above)
Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb]: I made this change

File: app/jobs/warm_bgs_caches_job.rb (hunk 1 above)
Comment (rubaiyat22) [reply app/jobs/warm_bgs_caches_job.rb]: We're alreadying using the last_synced_at field on BgsPowerOfAttorneys model to cache the 1000 bgs poa records that synced [last](https://github.com/department-of-veterans-affairs/caseflow/blob/fafdf207a62977510b3e5a2a7ce542885f8f6edc/app/jobs/warm_bgs_caches_job.rb#L114). For this [method](https://github.com/department-of-veterans-affairs/caseflow/blob/fafdf207a62977510b3e5a2a7ce542885f8f6edc/app/jobs/warm_bgs_caches_job.rb#L107), we could change it to so that i grabs claimants that don't have bgs poa records and sync those

File: app/jobs/warm_bgs_caches_job.rb [hunk 9]
```diff
@@ -172,26 +174,28 @@ def fetch_bgs_power_of_attorney_by_file_number(file_number, appeal_id)
...
   rescue ActiveRecord::RecordInvalid # not found at BGS
     BgsPowerOfAttorney.new(file_number: file_number)
-  rescue Errno::ECONNRESET, Savon::HTTPError => error
-    warning_msgs << "#{LegacyAppeal.name} #{appeal_id}: #{error}" if warning_msgs.count < 100
+  rescue StandardError => error
+    add_warning_msg("#{LegacyAppeal.name} #{appeal_id}: #{error}")
     nil
   end
 
   def warm_bgs_poa_and_return_cache_data(bgs_poa, appeal_id, appeal_type)
```
Comment (yoomlam) [inline app/jobs/warm_bgs_caches_job.rb:182]: Wondering if this method should be renamed since it originally returned a hash?

File: app/jobs/warm_bgs_caches_job.rb [hunk 10]
```diff
@@ -212,13 +216,20 @@ def claimants_for_open_appeals
...
 
   def oldest_claimants_for_open_appeals
     claimants_for_open_appeals.order(updated_at: :asc)
   end
 
+  def claimant_poa_or_nil(claimant)
+    claimant.power_of_attorney
+  rescue StandardError => error
+    add_warning_msg("#{Appeal.name} #{claimant.decision_review_id}: #{error}")
+    nil # returning nil here to allow job to continue; this does not mean that claimant is missing POA
```
Comment (yoomlam) [inline app/jobs/warm_bgs_caches_job.rb:230]: I appreciate the code comment!

**Review by yoomlam — APPROVED** [review]
LGTM  🚀
